_rules/table-headers-attribute-refer-to-data-cells-a25f45.md

```diff
@@ -0,0 +1,295 @@
+- [H43: Using id and headers attributes to associate data cells with header cells in data tables](https://www.w3.org/WAI/WCAG21/Techniques/html/H43)
+- [F90: Incorrectly associating table headers and content via the headers and id attributes](https://www.w3.org/WAI/WCAG21/Techniques/failures/F90)
+
+## Test Cases
+
+### Passed
+
+#### Passed Example 1
+
+The `headers` attribute on the cell refers to a `th` element within the same `table`.
+
+```html
+<table>
+	<thead>
+		<tr>
+			<th id="header1">Projects</th>
+		</tr>
+	</thead>
+	<tbody>
+		<tr>
+			<td headers="header1">15%</td>
+		</tr>
+	</tbody>
+</table>
+```
+
+#### Passed Example 2
+
+The `headers` attribute on the cell refers to a `th` element within the same `table`. Multiple headers are referenced for a cell with `colspan` of `2`.
+
+```html
+<table>
+	<thead>
+		<tr>
+			<th id="header1">Projects</th>
+			<th id="header2">Exams</th>
+		</tr>
+	</thead>
+	<tbody>
+		<tr>
+			<td colspan="2" headers="header1 header2">15%</td>
+		</tr>
+	</tbody>
+</table>
+```
+
+#### Passed Example 3
+
+The `headers` attribute on the second cell refers to a `td` element with a role of `columnheader` within the same `table`.
+
+```html
+<table>
+	<thead>
+		<tr>
+			<td role="columnheader" id="header1">Projects</td>
+		</tr>
+	</thead>
+	<tbody>
+		<tr>
+			<td headers="header1">15%</td>
+		</tr>
+	</tbody>
+</table>
+```
+
+#### Passed Example 4
+
+This `table` has multiple elements with a role of `columnheader`. The `headers` attribute on the cells lists IDs of `th` elements within the same `table`.
+
+```html
+<table>
+	<tr>
+		<th colspan="2" id="header1">Projects</th>
+		<th colspan="2" id="header2">Exams</th>
+	</tr>
+	<tr>
+		<th id="e1" headers="header1">1</th>
+		<th id="e2" headers="header1">2</th>
+		<th id="p1" headers="header2">1</th>
+		<th id="p2" headers="header2">2</th>
+	</tr>
+	<tr>
+		<td colspan="2" headers="header1 e1 e2">15%</td>
+		<td headers="header2 p1">15%</td>
+		<td headers="header2 p2">45%</td>
+	</tr>
+</table>
+```
+
+#### Passed Example 5
+
+The `headers` attribute on the cell refers to a `th` element with a role of `rowheader` within the same `table`.
+
+```html
+<table>
+	<tbody>
+		<tr>
+			<th role="rowheader" id="headerAge">Age</th>
+			<td headers="headerAge">65</td>
+		</tr>
+	</tbody>
+</table>
+```
+
+#### Passed Example 6
+
+The `headers` attribute on the last two `th` elements refers to another `th` element within the same `table`. Here the column header has a span of two columns.
+
+```html
+<table>
+	<tr>
+		<th id="name" colspan="2">Name</th>
+	</tr>
+	<tr>
+		<th headers="name">Firsname</th>
+		<th headers="name">Lastname</th>
+	</tr>
+</table>
+```
+
+### Failed
+
+#### Failed Example 1
+
+The `headers` attribute on the cell refers to an element outside the same `table`.
+
+```html
+<span id="elmOutsideTable">Project Costs</span>
+
+<table>
+	<tr>
+		<th id="headerOfColumn">Projects</th>
+	</tr>
+	<tr>
+		<td headers="elmOutsideTable">15%</td>
```

Comment:
This comment was never addressed: https://github.com/act-rules/act-rules.github.io/pull/412#discussion_r318026911 Like I said before, I don't think failed example 1 and 2 actually fail the success criterion.

Reply:
I disagree, sort of. Reading 1.3.1 is seems to be about the programatic information conveying the same association that is conveyed visually.

Failed examples 1 and 2 certainly match the rule expectations.

For failed example 1, it could be more clear that the span does not visually head up the table cell. For failed example 2, it could be that the Id doesn't exist in the page at all.

It strikes me that both examples are maybe an `aria-labelledby` away from not being a fail example. A point of view that raises questions about the rule and the referenced SC technique possibly not being detailed enough.

Reply:
This needs a little research.

Reply:
As per the [Algorithm for assigning header cells](https://html.spec.whatwg.org/multipage/tables.html#algorithm-for-assigning-header-cells), the position of the cell is only looked at if there is no `headers` attribute.

So, I think that in this case there is no programmatic association. (maybe UAs and ATs are fallbacking to the position if the `headers` attribute is KO...)

Reply:
Every screen reader we've tested this in will look at row/column headers for fallback. That may not be what the algorithm says they should do, but someone testing with one of the top 5 AT / browser combinations is not going to see a problem here. I find this test case questionable. Here's the test case we ran a little while ago:

```html
<table>
  <tr>
    <th scope="col">Projects</th>
    <th scope="col">Progress</th>
  </tr>
  <tr>
    <td headers="projects100">My Project</td>
    <td>15%</td>
  </tr>
</table>
```
**IE + JAWS**: Works
**Chrome + JAWS**: Works
**Firefox + NVDA**: Works
**Safari + VO**: Works

Reply:
If this passes the SC (which I don't challenge), then the rule need to be changed to take fallback headers into account.
This certainly fails the rule right now,  so if this passes the SC, the rule is bad.

And the rule in its current version is not referencing the assignation algorithm anymore…

(technically, the rule does assume that there is no fallback headers, but if this assumption is false for the vaste majority of cases, it needs to go away)